OK @dellis1972 , this is working as expect now with one little "migration" issue.

After updating to the nugets included here, the projects are incompatible with the existing "Target" definition in the csproj.

```
  <Target Name="RestoreDotnetTools" BeforeTargets="CollectPackageReferences">
    <Message Text="Restoring dotnet tools (this might take a while depending on your internet speed and should only happen upon building your project for the first time, or after upgrading MonoGame, or clearing your nuget cache)" Importance="High" />
    <Exec Command="dotnet tool restore" />
  </Target>
```

If the above exists in the project, it will not compile and results in the following error (for all project types)

```
 error MSB3073: The command "dotnet tool restore" exited with code 1.
```

Is this intended?  We can implement a migration guide, but this will likely be painful, as this is essentially a breaking change for all existing projects.

--- a/Tools/MonoGame.Content.Builder.Task/MonoGame.Content.Builder.Task.props
+++ b/Tools/MonoGame.Content.Builder.Task/MonoGame.Content.Builder.Task.props
@@ -8,7 +8,6 @@
     <AutoRestoreMGCBTool Condition="'$(AutoRestoreMGCBTool)' == '' And '$(MGCBCommand)' != ''">false</AutoRestoreMGCBTool>
     <AutoRestoreMGCBTool Condition="'$(AutoRestoreMGCBTool)' == ''">true</AutoRestoreMGCBTool>
     <MGCBCommand Condition="'$(MGCBCommand)' == ''">mgcb</MGCBCommand>
-    <MonoGameVersion Condition="'$(MonoGameVersion)' == ''">3.8.4.0</MonoGameVersion>
   </PropertyGroup>
 
 </Project>

--- a/Tools/MonoGame.Content.Builder.Task/MonoGame.Content.Builder.Task.props
+++ b/Tools/MonoGame.Content.Builder.Task/MonoGame.Content.Builder.Task.props
@@ -7,8 +7,6 @@
     <!-- If the user is using a custom $(MGCBCommand) , disable auto restore. -->
     <AutoRestoreMGCBTool Condition="'$(AutoRestoreMGCBTool)' == '' And '$(MGCBCommand)' != ''">false</AutoRestoreMGCBTool>
     <AutoRestoreMGCBTool Condition="'$(AutoRestoreMGCBTool)' == ''">true</AutoRestoreMGCBTool>
-    <MGCBToolDirectory Condition="'$(MGCBToolDirectory)' == ''">$(MSBuildThisFileDirectory)dotnet-tools/</MGCBToolDirectory>
-    <MGCBToolDirectory Condition="'$(MGCBToolDirectory)' != '' And !HasTrailingSlash('$(MGCBToolDirectory)')">$(MGCBToolDirectory)/</MGCBToolDirectory>
     <MGCBCommand Condition="'$(MGCBCommand)' == ''">mgcb</MGCBCommand>
     <MonoGameVersion Condition="'$(MonoGameVersion)' == ''">3.8.4.0</MonoGameVersion>
   </PropertyGroup>

--- a/Tools/MonoGame.Content.Builder.Task/MonoGame.Content.Builder.Task.targets
+++ b/Tools/MonoGame.Content.Builder.Task/MonoGame.Content.Builder.Task.targets
@@ -84,15 +84,6 @@
 
   </Target>
 
-  <!-- Restore the dotnet-mgcb tool to a known location. -->
-  <!-- Use MGCBToolAdditionalArguments to provide additional arguments to the install
-   for example a path to your custom NuGet.config file.
-  -->
-  <Target Name="RestoreContentCompiler" Condition=" '$(AutoRestoreMGCBTool)' == 'true' And !Exists ('$(MGCBToolDirectory)$(MGCBCommand)')">
-    <MakeDir Directories="$(MGCBToolDirectory)"/>
-    <Exec Command="&quot;$(DotnetCommand)&quot; tool install $(MGCBToolAdditionalArguments) dotnet-mgcb --version $(MonoGameVersion) --tool-path ." WorkingDirectory="$(MGCBToolDirectory)" ContinueOnError="true" />
-  </Target>
-
   <!--
     =====================
     PrepareContentBuilder
@@ -132,6 +123,21 @@
 
   </Target>
 
+  <Target Name="_RestoreMGCBTool" Condition="'$(AutoRestoreMGCBTool)' == 'true'"
+        Inputs="$(MSBuildProjectFullPath)"
+        Outputs="$(IntermediateOutputPath)mgcb.tool.restore">
+    <Exec
+      Condition="'$(DotnetCommand)' == ''"
+      Command="&quot;$(DotnetCommand)&quot; tool restore"
+      WorkingDirectory="$(MSBuildProjectDirectory)"
+      ContinueOnError="true"
+    />
+    <Touch Files="$(IntermediateOutputPath)mgcb.tool.restore" AlwaysCreate="true" />
+    <ItemGroup>
+      <FileWrites Include="$(IntermediateOutputPath)mgcb.tool.restore" />
+    </ItemGroup>
+  </Target>
+
   <!--
     =================
     RunContentBuilder
@@ -143,11 +149,9 @@
       - ExtraContent: built content files
         - ContentDir: the relative path of the embedded folder to contain the content files
   -->
-  <Target Name="RunContentBuilder" DependsOnTargets="RestoreContentCompiler;PrepareContentBuilder">
+  <Target Name="RunContentBuilder" DependsOnTargets="_RestoreMGCBTool;PrepareContentBuilder">
 
     <PropertyGroup>
-      <_Command Condition="Exists ('$(MGCBToolDirectory)$(MGCBCommand)')">&quot;$(MGCBToolDirectory)$(MGCBCommand)&quot;</_Command>
-      <!-- Fallback to old behaviour this allows people to override $(MGCBCommand) with the mgcb.dll -->
       <_Command Condition=" '$(_Command)' == '' ">&quot;$(DotnetCommand)&quot; &quot;$(MGCBCommand)&quot;</_Command>
     </PropertyGroup>
 
@@ -173,6 +177,8 @@
     ==============
 
     Include ExtraContent as platform-specific content in the project output.
+    Runs just before the complile step. But after the project references have been resolved.
+    This allows for content pipeline extensions to be built first.
 
     Outputs:
       - AndroidAsset: built content files if Android
@@ -184,7 +190,8 @@
     DependsOnTargets="RunContentBuilder"
     Condition="'$(EnableMGCBItems)' == 'true' OR '@(MonoGameContentReference)' != ''"
     Outputs="%(ExtraContent.RecursiveDir)%(ExtraContent.Filename)%(ExtraContent.Extension)"
-    BeforeTargets="BeforeBuild">
+    BeforeTargets="BeforeCompile"
+    AfterTargets="ResolveProjectReferences">
 
     <!-- Call CreateItem on each piece of ExtraContent so it's easy to switch the item type by platform. -->
     <CreateItem

--- a/Tools/MonoGame.Content.Builder.Task/MonoGame.Content.Builder.Task.targets
+++ b/Tools/MonoGame.Content.Builder.Task/MonoGame.Content.Builder.Task.targets
@@ -123,10 +123,18 @@
 
   </Target>
 
+  <Target Name="_CollectDotnetToolConfigFiles">
+    <ItemGroup>
+      <_DotnetToolConfigFiles Include="$(SolutionDir)/.config/dotnet-tools.json" Condition="Exists('$(SolutionDir)/.config/dotnet-tools.json')" />
+      <_DotnetToolConfigFiles Include="$(MSBuildProjectDirectory)/.config/dotnet-tools.json" Condition="Exists('$(MSBuildProjectDirectory)/.config/dotnet-tools.json')" />
+    </ItemGroup>
+  </Target>
+
   <Target Name="_RestoreMGCBTool"
         Condition="'$(AutoRestoreMGCBTool)' == 'true'"
         BeforeTargets="CollectPackageReferences"
-        Inputs="$(MSBuildProjectFullPath)"
+        DependsOnTargets="_CollectDotnetToolConfigFiles"
+        Inputs="$(MSBuildProjectFullPath);@(_DotnetToolConfigFiles);$(ProjectAssetsFile)"
         Outputs="$(IntermediateOutputPath)mgcb.tool.restore">
     <Exec
       Condition="'$(DotnetCommand)' != ''"

--- a/Tools/MonoGame.Content.Builder.Task/MonoGame.Content.Builder.Task.targets
+++ b/Tools/MonoGame.Content.Builder.Task/MonoGame.Content.Builder.Task.targets
@@ -151,7 +151,7 @@
       - ExtraContent: built content files
         - ContentDir: the relative path of the embedded folder to contain the content files
   -->
-  <Target Name="RunContentBuilder" DependsOnTargets="_RestoreMGCBTool;PrepareContentBuilder">
+  <Target Name="RunContentBuilder" DependsOnTargets="PrepareContentBuilder">
 
     <PropertyGroup>
       <_Command Condition=" '$(_Command)' == '' ">&quot;$(DotnetCommand)&quot; &quot;$(MGCBCommand)&quot;</_Command>

